mlflow/sagemaker/__init__.py

```diff
@@ -316,36 +385,128 @@ def _upload_s3(local_model_path, bucket, prefix):
     """
     sage_client = boto3.client('sagemaker', region_name=region_name)
-    model_name = app_name + '-model'
+    s3_client = boto3.client('s3', region_name=region_name)
+
+    deployed_endpoints_page = sage_client.list_endpoints(MaxResults=100)
+    endpoint_found = (app_name in [endp["EndpointName"] for endp in deployed_endpoints_page["Endpoints"]])
+    while (not endpoint_found) and ("NextToken" in deployed_endpoints_page):
+        next_token = deployed_endpoints_page["NextToken"]
+        deployed_endpoints_page = sage_client.list_endpoints(MaxResults=100, NextToken=next_token)
+        for endp in deployed_endpoints_page["Endpoints"]:
+            if endp["EndpointName"] == app_name:
+                endpoint_found = True
+                break
+
+    if endpoint_found and mode == DEPLOYMENT_MODE_CREATE:
+        msg = ("You are attempting to deploy an application with name: `{an}` in `{mcr} `mode." 
+               " However, an application with the same name already exists. If you want to update" 
+               " this application, deploy in `{madd}` or `{mrep}` mode.").format(
+                       an=app_name,
+                       mcr=DEPLOYMENT_MODE_CREATE,
+                       madd=DEPLOYMENT_MODE_ADD,
+                       mrep=DEPLOYMENT_MODE_REPLACE)
+        
+        raise Exception(msg)
+    
+    elif endpoint_found:
+        return _update_sagemaker_endpoint(endpoint_name=app_name,
+                                          image_url=image_url,
+                                          model_s3_path=model_s3_path,
+                                          run_id=run_id,
+                                          mode=mode, 
+                                          archive=archive,
+                                          role=role,
+                                          sage_client=sage_client,
+                                          s3_client=s3_client)
+
+    else:
+        return _create_sagemaker_endpoint(endpoint_name=app_name,
+                                          image_url=image_url,
+                                          model_s3_path=model_s3_path,
+                                          run_id=run_id,
+                                          role=role,
+                                          sage_client=sage_client,
+                                          s3_client=s3_client)
+
+
+def _get_sagemaker_resource_unique_id():
+    """
+    :return: A unique, timestamp-based identifier that can be
+    appended to a user-readable resource name to avoid naming
+    collisions.
+    """
+    dt = datetime.now()
+    epoch = datetime.utcfromtimestamp(0)
+    # Our unique identifier uses milliseconds-since-epoch to avoid errant
+    # behavior in the case where multiple users attempt to deploy models
+    # to the same endpoint at roughly the same time
+    delta_millis = int((dt - epoch).total_seconds() * 1000)
```

Comment:
Ah actually thinking about it, I feel we should use UUIDs (can generate with `uuid.uuid4().hex`) instead of timestamps - this would just avoid any concerns about the odds of a collision increasing with the number of concurrent users in an AWS account & be consistent with how we uniquely identify other objects in MLflow. The only downside I see is that the names are harder to read, but it seems a user using the SageMaker CLI to list endpoints / models would see endpoints / models sorted by creation time anyways (see [CLI docs](https://docs.aws.amazon.com/cli/latest/reference/sagemaker/list-endpoints.html)). Sorry for the flip-flop & let me know what you think!

Reply:
While the UUID definitely solves the user collision problem, it doesn't help to avoid naming conflicts when a user creates multiple objects associated with the same application. For example, during the `update` procedure for an endpoint, we create a new endpoint configuration object and deploy it before deleting the old one (this order is used to ensure that a creation failure doesn't leave the endpoint without a configuration). These two configurations cannot share the same name. One option is to use a combination of UUID and a timestamp; however, this eats up a significant portion of the 63-character maximum length of a SageMaker model / configuration resource name.

A millisecond-resolution timestamp seems attractive because it significantly reduces the odds of collision in the multi-user case, avoids object naming conflicts within a single application, and doesn't eat up too much of the naming bandwidth. Please let me know what your thoughts are.

Reply:
Based on offline discussion with @smurching , I was confusing the term UUID (universally unique identifier) with UDID (unique device identifier). We've decided to use a base64-encoded UUID to cut down on the length of the identifier while maintaining uniqueness.